parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -256,23 +260,53 @@ type TermBufferEntry = BufferEntry (Term Symbol Ann, Type Symbol Ann)
 
 type DeclBufferEntry = BufferEntry (Decl Symbol Ann)
 
-unsafeGetConnection :: MonadIO m => Codebase.DebugName -> CodebasePath -> m Connection
+-- | Create a new sqlite connection to the database at the given path.
+--   the caller is responsible for calling the returned cleanup method once finished with the
+--   connection.
+--   The connection may not be used after it has been cleaned up.
+--   Prefer using 'withConnection' if you can, as it guarantees the connection will be properly
+--   closed for you.
+unsafeGetConnection ::
+  MonadIO m =>
+  Codebase.DebugName ->
+  CodebasePath ->
+  m (IO (), Connection)
 unsafeGetConnection name root = do
   let path = root </> codebasePath
   Monad.when debug $ traceM $ "unsafeGetconnection " ++ name ++ " " ++ root ++ " -> " ++ path
   (Connection name path -> conn) <- liftIO $ Sqlite.open path
   runReaderT Q.setFlags conn
-  pure conn
-
-shutdownConnection :: MonadIO m => Connection -> m ()
-shutdownConnection conn = do
-  Monad.when debug $ traceM $ "shutdown connection " ++ show conn
-  liftIO $ Sqlite.close (Connection.underlying conn)
-
-sqliteCodebase :: (MonadIO m, MonadCatch m) => Codebase.DebugName -> CodebasePath -> m (Either SchemaVersion (m (), Codebase m Symbol Ann))
+  -- Closing the connection more than once is a no-op
+  cleanup <- liftIO $ once (shutdownConnection conn)
```

Comment:
Small nit: I'm not so sure `once` is desirable here. 

On one hand, I really like making APIs easier to use correctly. And if that means we can eliminate or relax tricky conditions like, "you _must_ call this function _exactly once_", then that's great.

However, in this particular case, a connection is already a stateful IO-ey thing; closing it will make it busted for all other threads that are happening to use it. So really, the opening/closing of the connection is already the responsibility of exactly one thread. (Our connection also happens to not be thread safe at the moment anyway).

Given that, I think it is appropriate for `close` to throw an exception if called twice, to alert us of a bug, which `once` would silence.

Reply:
I thought about that too, I'm not extremely convinced either way.
I added this because in a few spots in tests it was easy for folks to write code where their exception path overlapped with their "happy" path, and they would close the connection twice.

I think it's probably fine, since if there really is a bug we'll still get the exception from trying to run queries on a dead connection anyways right?

Reply:
I'm thinking if we can catch a future bug by throwing an exception here before a user encounters it by trying to run queries on a dead connection, we should do that.  But weighing against the effort of fixing up the tests, I'm not sure how to decide either.

How much work would it be to get tests updated?

In either case, let's document the decision in the code, since it'll be easier to find than in the PR review comments.